### PR TITLE
ConfigurationGathering.py: Irrelevant warning

### PR DESCRIPTION
**Pull request by incorrectusername**
While running `coala -B` or `coala -B -l <language>` ,
if the .coafile is not found then coala prints
irrelevent warnings.

Fixes coala#2775

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [ ] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [ ] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by incorrectusername** [comment]
http://pastebin.com/3cWi3jMD . getting StyleLintBear warning don't know why.

**Comment by adhikasp** [comment]
Looks fine, try to repush again?

**Comment by Udayan12167** [comment]
Tests using that argument please. The goal is not to get 100% coverage but to actually test the functionality you have added 😄

**Comment by incorrectusername** [comment]
@Udayan12167 i have tested using `coala -B -l c` :) . can you please tell me again what needs to be done 😅 :)

**Comment by RohanVB** [comment]
@incorrectusername I think he meant you have to write the corresponding unittests in `ConfigurationGatheringTest.py`

**Comment by incorrectusername** [comment]
I need help with the tests. 
i cannot understand what to do if i call `get_filtered_bears` from `ConfigurationGatheringTest.py` then what should i do with the tuple returned.
also which assert to use after that.

if my approach is wrong then please direct me.

**Comment by incorrectusername** [comment]
Can anyone help me with the tests?
@Udayan12167 ?

File: coalib/settings/ConfigurationGathering.py [hunk 1]
```diff
@@ -136,7 +136,7 @@ def warn_config_absent(sections, argument, log_printer):
     return False
 
 
-def load_configuration(arg_list, log_printer, arg_parser=None):
+def load_configuration(arg_list, log_printer, arg_parser=None, silent=False):
     """
     Parses the CLI args and loads the config file accordingly, taking
     default_coafile and the users .coarc into account.
```
Comment (Nosferatul) [inline coalib/settings/ConfigurationGathering.py]: Explain what is `silent` between `log_printer` and `return`

Comment (incorrectusername) [inline coalib/settings/ConfigurationGathering.py]: The message `The default coafile '.coafile' was not found.......` we get when the default coafile was not found( `load_config_file` is responsible for that. ) . so if `silent = True` then that statement is not executed under `load_config_file` . And when we are just checking for bears we don't need coafile . 

So while calling `load_configuration` from `get_filtered_bears` if we pass `silent=True` with `load_configuration` then we can use this `silent` to not warn. 

hope this helps :) . 

Can you help me with tests? :)

File: coalib/settings/ConfigurationGathering.py [hunk 2]
```diff
@@ -170,7 +170,8 @@ def load_configuration(arg_list, log_printer, arg_parser=None):
             log_printer,
             silent=True)
 
-        default_config = str(base_sections['default'].get('config', '.coafile'))
+        default_config = str(
+            base_sections['default'].get('config', '.coafile'))
```
Comment (aptrishu) [inline coalib/settings/ConfigurationGathering.py]: unrelated change ?

File: tests/settings/ConfigurationGatheringTest.py [hunk 3]
```diff
@@ -333,3 +333,21 @@ def test_default_section_deprecation_warning(self):
...
             self.assertNotIn('WARNING', stdout.getvalue())
+
+    def test_show_bears(self):
+        current_dir = os.path.abspath(os.path.dirname(__file__))
+        child_dir = os.path.join(current_dir,
+                                 'section_manager_test_files',
+                                 'child_dir')
+        with change_directory(child_dir):
+            with self.assertRaises(SystemExit) as cm:
+                sections, target = load_configuration(
```
Comment (aptrishu) [inline tests/settings/ConfigurationGatheringTest.py]: verify the exact exception, use `assertIn` to make things simple.

File: coalib/settings/ConfigurationGathering.py [hunk 4]
```diff
@@ -136,7 +136,7 @@ def warn_config_absent(sections, argument, log_printer):
     return False
 
 
-def load_configuration(arg_list, log_printer, arg_parser=None):
+def load_configuration(arg_list, log_printer, arg_parser=None, silent=False):
     """
```
Comment (aptrishu) [inline coalib/settings/ConfigurationGathering.py]: Update the `:param:` docstring please.

File: coalib/settings/ConfigurationGathering.py (hunk 2 above)
Comment (incorrectusername) [inline coalib/settings/ConfigurationGathering.py]: it was PEP8Bear :P

Comment (jayvdb) [inline coalib/settings/ConfigurationGathering.py]: Try again please. This line passes CI in builds every day

File: tests/settings/ConfigurationGatheringTest.py [hunk 5]
```diff
@@ -14,7 +14,8 @@
 from coalib.output.printers.LogPrinter import LogPrinter
 from coala_utils.string_processing import escape
 from coalib.settings.ConfigurationGathering import (
-    find_user_config, gather_configuration, load_configuration)
+    find_user_config, gather_configuration, load_configuration,
```
Comment (jayvdb) [inline tests/settings/ConfigurationGatheringTest.py]: One per line, in alpha order, with trailing comma and trailing round bracket on a separate line.

File: coalib/settings/ConfigurationGathering.py [hunk 6]
```diff
@@ -136,13 +136,15 @@ def warn_config_absent(sections, argument, log_printer):
...
     Parses the CLI args and loads the config file accordingly, taking
     default_coafile and the users .coarc into account.
 
     :param arg_list:    The list of command line arguments.
     :param log_printer: The LogPrinter object for logging.
+    :param silent:      Whether or not to warn the user/exit if the file
+                        doesn't exist.
```
Comment (jayvdb) [inline coalib/settings/ConfigurationGathering.py]: This should mention it is ignored is `save` is enabled.

**Comment by incorrectusername** [comment]
The tests are failing, don't know why `SystemExit` doesn't get raised .can anyone help.

File: coalib/settings/ConfigurationGathering.py [hunk 7]
```diff
@@ -183,7 +186,11 @@ def load_configuration(arg_list, log_printer, arg_parser=None):
             # but to a specific file.
             save = True
 
-        coafile_sections = load_config_file(config, log_printer, silent=save)
+        if silent:
+            coafile_sections = load_config_file(config, log_printer, silent)
+        else:
+            coafile_sections = load_config_file(
+                config, log_printer, silent=save)
```
Comment (Makman2) [inline coalib/settings/ConfigurationGathering.py]: deduplicate code:
```
coafile_sections = load_config_file(config, log_printer, silent or save)
```

File: coalib/settings/ConfigurationGathering.py [hunk 8]
```diff
@@ -136,13 +136,15 @@ def warn_config_absent(sections, argument, log_printer):
...
-def load_configuration(arg_list, log_printer, arg_parser=None):
+def load_configuration(arg_list, log_printer, arg_parser=None, silent=False):
     """
     Parses the CLI args and loads the config file accordingly, taking
     default_coafile and the users .coarc into account.
 
     :param arg_list:    The list of command line arguments.
     :param log_printer: The LogPrinter object for logging.
+    :param silent:      Whether or not to warn the user/exit if the file
+                        doesn't exist.It is ignored when `save` is enabled.
```
Comment (Makman2) [inline coalib/settings/ConfigurationGathering.py]: Missing space after period

File: coalib/settings/ConfigurationGathering.py (hunk 2 above)
Comment (Makman2) [inline coalib/settings/ConfigurationGathering.py]: definitely unrelated change, should be omitted

File: tests/settings/ConfigurationGatheringTest.py [hunk 9]
```diff
@@ -14,7 +14,11 @@
 from coalib.output.printers.LogPrinter import LogPrinter
 from coala_utils.string_processing import escape
 from coalib.settings.ConfigurationGathering import (
-    find_user_config, gather_configuration, load_configuration)
+    find_user_config,
+    gather_configuration,
+    get_filtered_bears,
+    load_configuration,
+    )
```
Comment (Makman2) [inline tests/settings/ConfigurationGatheringTest.py]: rather like this, as this is our current style :)
```

```
If you dislike this, feel free to file an issue. This style is definitely worth thinking about if you believe that we should have this (because of VCS reasons) :)

**Comment by sils** [comment]
this commit message no verb :)

**Comment by sils** [comment]
maybe "Silence irrelevant warning"

**Comment by NiklasMM** [comment]
Ok, I rebased this but the tests are failing.

**Comment by incorrectusername** [comment]
Sorry, I have been little busy with college. 
what should I do next in order to get it merged? I don't have idea

**Comment by NiklasMM** [comment]
Rebase again and see why the tests are failing (if they still are after rebasing) :+1:

**Comment by NiklasMM** [comment]
`ConfigurationGatheringCollectionTest.test_show_bears` is failing.

**Comment by satwikkansal** [comment]
@incorrectusername around?

I've made a similar PR #4622 being unaware that your PR already exists for the similar issue. If you're little busy till the next weekend, I can take up this issue (as it is kinda required for my GSoC project :sweat_smile: )

**Comment by incorrectusername** [comment]
Hi @satwikkansal , This issue is already solved. Only the tests are failing. Isaw your PR if this doesn't require any tests than you can merge this.

**Comment by satwikkansal** [comment]
@incorrectusername okay, I combine your and mine desired changes, added tests, and updated the PR #4622

**Comment by satwikkansal** [comment]
Okay #4622 is now merged, @incorrectusername would you like to close this PR now?

PS: Thanks for the coordination
